File: identhub/src/main/java/de/solarisbank/identhub/verfication/phone/VerificationPhoneViewModel.java
package de.solarisbank.identhub.verfication.phone;

import android.os.CountDownTimer;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.util.Locale;
import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.disposables.CompositeDisposable;

public class VerificationPhoneViewModel extends ViewModel {

    public static final int DEFAULT_STATE = 0;
    public static final int SUCCESS_STATE = 1;
    public static final int ERROR_STATE = 2;
    public static final int DISABLED_STATE = 3;
    private static final int ONE_SEC = 1;
    private static final int COUNTER_TIME = 20;
    private static final int MIN_CODE_LENGTH = 6;
    private final CompositeDisposable compositeDisposable;
    private final MutableLiveData<CountDownTimeState> timerLiveData;
    private final CountDownTimer countDownTimer;
    private MutableLiveData<Integer> state;

    public VerificationPhoneViewModel() {
        compositeDisposable = new CompositeDisposable();
        timerLiveData = new MutableLiveData<>();
        countDownTimer = new CountDownTimer(TimeUnit.SECONDS.toMillis(COUNTER_TIME), TimeUnit.SECONDS.toMillis(ONE_SEC)) {
            @Override
            public void onTick(long millisUntilFinished) {
                String currentTime = String.format(Locale.getDefault(), "%02d:%02d",
                        TimeUnit.MILLISECONDS.toMinutes(millisUntilFinished) % 60,
                        TimeUnit.MILLISECONDS.toSeconds(millisUntilFinished) % 60);
                timerLiveData.postValue(new CountDownTimeState(currentTime));
            }

            @Override
            public void onFinish() {
                CountDownTimeState currentState = timerLiveData.getValue();
                timerLiveData.postValue(new CountDownTimeState(currentState.getCurrentValue(), true));
            }
        };

    }

    public LiveData<Integer> getState() {
        if (state == null) {
            state = new MutableLiveData<>(0);
        }

        return state;
    }

    public LiveData<CountDownTimeState> getTimer() {
        return timerLiveData;
    }

    public void onSubmitButtonClicked() {
        int state = this.state.getValue();

        state = state == DEFAULT_STATE ? ERROR_STATE : DEFAULT_STATE;

        if (state == DEFAULT_STATE) {
            startTimer();
        } else {
            stopTimer();
        }

        this.state.postValue(state);
    }


    public void onSendNewCodeClicked() {
        state.postValue(DEFAULT_STATE);
        startTimer();
    }

    public void startTimer() {
        countDownTimer.start();
    }

    public void stopTimer() {
        countDownTimer.cancel();
    }

    public void onVerificationCodeReady(String code) {
        if (code.length() == MIN_CODE_LENGTH) {
            stopTimer();
            state.postValue(DISABLED_STATE);
            compositeDisposable.add(
                    Observable.just(SUCCESS_STATE)
                            .delay(6, TimeUnit.SECONDS)
                            .subscribe(integer -> state.postValue(SUCCESS_STATE))
            );
        }
    }

    @Override
    protected void onCleared() {
        compositeDisposable.clear();
    }
}

File: identhub/src/main/java/de/solarisbank/identhub/verfication/phone/VerificationPhoneViewModel.kt
package de.solarisbank.identhub.verfication.phone

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.jakewharton.rxrelay2.BehaviorRelay
import de.solarisbank.identhub.data.verification.phone.model.VerificationPhoneResponse
import de.solarisbank.identhub.domain.verification.phone.AuthorizeVerificationPhoneUseCase
import de.solarisbank.identhub.domain.verification.phone.ConfirmVerificationPhoneUseCase
import de.solarisbank.identhub.event.ClickEvent
import de.solarisbank.identhub.progress.DefaultCountDownTimer
import de.solarisbank.sdk.core.result.Event
import de.solarisbank.sdk.core.result.Result
import de.solarisbank.sdk.core.result.succeeded
import io.reactivex.disposables.CompositeDisposable
import java.util.concurrent.TimeUnit

class VerificationPhoneViewModel(
        private val authorizeVerificationPhoneUseCase: AuthorizeVerificationPhoneUseCase,
        private val confirmVerificationPhoneUseCase: ConfirmVerificationPhoneUseCase
) : ViewModel() {
    private val authorizeResultLiveData: MutableLiveData<Result<VerificationPhoneResponse>> = MutableLiveData()
    private val compositeDisposable: CompositeDisposable = CompositeDisposable()
    private val countDownTimeEventLiveData: MutableLiveData<Event<CountDownTime>> = MutableLiveData()
    private val clickEventRelay = BehaviorRelay.createDefault(ClickEvent())
    private var countDownTimer = DefaultCountDownTimer(TimeUnit.SECONDS.toMillis(COUNTER_TIME), TimeUnit.SECONDS.toMillis(INTERVAL_IN_SEC))
    private val confirmResultLiveData: MutableLiveData<Result<VerificationPhoneResponse>> = MutableLiveData()
    private val tickListener = object : DefaultCountDownTimer.OnTickListener {
        override fun onTick(millisUntilFinished: Long) {
            countDownTimeEventLiveData.postValue(Event(CountDownTime(millisUntilFinished)))
        }

        override fun onFinish(millisUntilFinished: Long) {
            countDownTimeEventLiveData.postValue(Event(CountDownTime(millisUntilFinished, true)))
        }
    }

    init {
        authorizeTanOnClick()
        countDownTimer.addListener(tickListener)
    }

    private fun authorizeTanOnClick() {
        compositeDisposable.add(
                clickEventRelay.doOnNext { authorizeResultLiveData.postValue(Result.Loading) }
                        .switchMap {
                            authorizeVerificationPhoneUseCase.execute(Unit)
                                    .toObservable()
                        }
                        .subscribe({ onAuthorizedReceived(it) }) { onAuthorizedError(it) })
    }

    private fun onAuthorizedError(throwable: Throwable) {
        stopTimer()
        authorizeResultLiveData.postValue(Result.createUnknown(throwable))
    }

    private fun onAuthorizedReceived(result: Result<VerificationPhoneResponse>) {
        if (result.succeeded) {
            startTimer()
        } else {
            stopTimer()
        }
        authorizeResultLiveData.postValue(result)
    }

    fun getAuthorizeResultLiveData(): LiveData<Result<VerificationPhoneResponse>> {
        return authorizeResultLiveData
    }

    fun getCountDownTimeEventLiveData(): LiveData<Event<CountDownTime>> {
        return countDownTimeEventLiveData
    }

    fun getConfirmResultLiveData(): LiveData<Result<VerificationPhoneResponse>> {
        return confirmResultLiveData
    }

    fun onSubmitButtonClicked(confirmToken: String) {
        stopTimer()
        confirmResultLiveData.postValue(Result.Loading)
        compositeDisposable.add(
                confirmVerificationPhoneUseCase.execute(confirmToken)
                        .subscribe(
                                { notifyConfirmResultChanged(it) },
                                { notifyConfirmResultChanged(Result.createUnknown(it)) }
                        )
        )
    }

    private fun notifyConfirmResultChanged(result: Result<VerificationPhoneResponse>) {
        confirmResultLiveData.postValue(result)
    }

    fun onSendNewCodeClicked() {
        clickEventRelay.accept(ClickEvent())
    }

    fun startTimer() {
        countDownTimer.start()
    }

    fun stopTimer() {
        countDownTimer.cancel()
    }

    override fun onCleared() {
        Log.d("addOnBackStackChanged", "onCleared " + javaClass.name)
        countDownTimer.removeListener(tickListener)
        countDownTimer.cancel()
        compositeDisposable.clear()
    }

    companion object {
        private const val INTERVAL_IN_SEC = 1L
        private const val COUNTER_TIME = 20L
        const val MIN_CODE_LENGTH = 6
    }
}